### PR TITLE
Add support for requirements handling with az notebook deploy

### PR DESCRIPTION
**Pull request by dzeih**

### CODE REVIEW

File: azure/databricks_notebook.py [hunk 1]
```diff
@@ -138,18 +142,158 @@ def inject_job_path(job_config: Dict[str, Any]) -> str:
...
+
+    url = f'{deploy_url}/libraries/install'
+    data = {'cluster_id': cluster_id, 'libraries': final_requirements}  # type: ignore
+    r = requests.post(url, headers=headers, json=data)
+    if r.status_code != 200:
+        raise Exception(f'Failure in requirements installation: {r.reason}')
+    print('Libs installed successfully')
+
+
+def create_requirements_install_info(deploy_url: str, requirements: Dict[str, Any]) -> List[Dict[str, Any]]:
```
Comment (latvanii) [inline azure/databricks_notebook.py]: Style hint for this function: put everything that is under if into separate function. If is always a good smell for a need for separate function.
This way this main function would be super readable, like:
```
def create_requirements_install_info(...)
    required_setup: List[Dict[str, Any]] = []
    if 'maven' in requirements:
        required_setup = maven_install_info(required_setup, ...)
    if 'pypi' in requirements:
        required_setup = pypi_install_info(required_setup, ...)
    ...
    return required_setup
```

File: azure/databricks_notebook.py [hunk 2]
```diff
@@ -138,18 +146,162 @@ def inject_job_path(job_config: Dict[str, Any]) -> str:
...
+    artifact_id = artifact_parts[1]
+
+    url = 'https://repository.sonatype.org/service/local/artifact/maven/redirect'
+    headers = {"content-type": "application/json"}
+    params = {"r": "central-proxy",
+              "g": group_id,
+              "a": artifact_id,
+              "v": "LATEST"}
+    r = call_get(url, headers=headers, params=params, error_message='Error in maven version check')
+    return r.url.split('/')[-2]
```
Comment (opah) [inline azure/databricks_notebook.py]: is there a need to check the url split separately and throw more context aware exception if expected parts are not present?

File: azure/databricks_notebook.py [hunk 3]
```diff
@@ -138,18 +146,162 @@ def inject_job_path(job_config: Dict[str, Any]) -> str:
...
+        existing_artifact_versions = [x['library']['maven']['coordinates'].split(':')
+                                      for x in previously_installed['library_statuses']
+                                      if 'maven' in x['library'] and x['status'] == 'INSTALLED']
+    for requirement in requirements:
+        # TODO: check local repos + pypi + jar versions somehow to avoid reinstalls if possible
+        if 'maven' in requirement:
+            coordinate_splitted = requirement['maven']['coordinates'].split(':')
+            is_installed_or_newer_exists = any(coordinate_splitted == existing or
+                                               (coordinate_splitted[2] < existing[2] and
+                                                coordinate_splitted[:2] == existing[:2])
```
Comment (opah) [inline azure/databricks_notebook.py]: Maybe put this a separate function whixch name desacribes what this does

File: azure/databricks_notebook.py [hunk 4]
```diff
@@ -138,18 +146,162 @@ def inject_job_path(job_config: Dict[str, Any]) -> str:
...
+            return True
+        return False
+
+    if not _is_cluster_running(cluster_url_prefix, data):
+        print('Cluster not running, starting')
+        r = call_post(url=f'{cluster_url_prefix}/start', headers=headers, json=data, error_message='Starting cluster')
+        print('Start command successful')
+        while not _is_cluster_running(cluster_url_prefix, data):
+            print('Waiting for cluster to start')
+            sleep(5)
```
Comment (opah) [inline azure/databricks_notebook.py]: This could be a separate function

File: azure/databricks_notebook.py [hunk 5]
```diff
@@ -138,18 +146,162 @@ def inject_job_path(job_config: Dict[str, Any]) -> str:
...
+        full_install_path = f"{target_install_path}/{current_head}_{egg_file}"
+        payload = {"path": full_install_path, "contents": convert_code_to_base64(f'{egg_path_prefix}/{egg_file}')}
+        r = call_post(url, headers=headers, json=payload, exceptions_not_allowed=False)
+        if r.status_code == 400 and r.json().get('error_code') == 'RESOURCE_ALREADY_EXISTS':
+            print(f'Current commit {current_head} already exists on the cluster, resuming without upload')
+        elif r.status_code != 200:
+            raise Exception(f'Error while uploading egg: {r.status_code}, {r.reason}, {r.text}')
+        required_repos.append({"egg": full_install_path})
+        assert (len(copy_path) > 0)
+        subproc_run(['rm', '-rf', copy_path])
```
Comment (opah) [inline azure/databricks_notebook.py]: this looks dangerous 😨

**Pull request merged**